vnext/Microsoft.ReactNative/JsiWriter.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "JsiWriter.h"
+#include <crash/verifyElseCrash.h>
+
+namespace winrt::Microsoft::ReactNative {
+
+//===========================================================================
+// JsiWriter implementation
+//===========================================================================
+
+JsiWriter::JsiWriter(facebook::jsi::Runtime &runtime) noexcept : m_runtime(runtime) {
+  Push({ContinuationAction::AcceptValueAndFinish});
+}
+
+facebook::jsi::Value JsiWriter::MoveResult() noexcept {
+  VerifyElseCrash(m_continuations.size() == 0);
+  return std::move(m_result);
+}
+
+facebook::jsi::Value JsiWriter::CopyResult() noexcept {
+  VerifyElseCrash(m_continuations.size() == 0);
+  return {m_runtime, m_result};
+}
+
+void JsiWriter::WriteNull() noexcept {
+  WriteValue(facebook::jsi::Value::null());
+}
+
+void JsiWriter::WriteBoolean(bool value) noexcept {
+  WriteValue({value});
+}
+
+void JsiWriter::WriteInt64(int64_t value) noexcept {
+  // JavaScript's integer is not int64_t, need to ensure that the value is in range
+  VerifyElseCrash(MinSafeInteger <= value && value <= MaxSafeInteger);
+
+  // unfortunately JSI only supports int and double for number, choose double here
+  // make sure that doing type conversion in C++ makes no data loss
+  double d = static_cast<double>(value);
+  VerifyElseCrash(floor(d) == d);
+  WriteValue(d);
+}
+
+void JsiWriter::WriteDouble(double value) noexcept {
+  WriteValue({value});
+}
+
+void JsiWriter::WriteString(const winrt::hstring &value) noexcept {
+  WriteValue({m_runtime, facebook::jsi::String::createFromUtf8(m_runtime, winrt::to_string(value))});
```

Comment:
>createFromUtf8 [](start = 48, length = 14)

I wonder: do they have a method to create from UTF16?

Reply:
I didn't find anything like that, they have ansi and utf8